### ISSUE TITLE
Flatpak downloads every nvidia driver from flathub at every update

### ISSUE THREAD

**Issue by trialism** [flathub/org.freedesktop.Platform.GL.nvidia#42]
Issue moved from here: https://github.com/flatpak/flatpak/issues/3750

## Linux distribution and version

Ubuntu 20.04

## Flatpak version

1.6.3

## Description of the problem

Every time a new nvidia driver gets released I need to update the flatpaks and this already slow process(with SSD and very fast internet) becomes slower because it redownloads every driver again. This happened three times so far. Now it downloaded every driver from 430.xy to 450.xy.

## Steps to reproduce

Run `flatpak update` when there is a new nvidia driver available.

**Comment by TingPing** [comment]
So currently the `build.sh` script simply builds every version every time.

What we need is for it to somehow detect if only a new version was added and only export those.

**Comment by 404NetworkError** [comment]
I've thought about this before. A potential problem I see is what would happen if URLs for older releases change or get removed. (It hasn't happened yet but could in the future.) By building new releases on each commit, it guarantees that the older releases still exist. There are definitely better ways to combat this problem, though.

**Comment by TingPing** [comment]
There are potentially really fancy things would could do in coordination with the buildbot, like maybe indicate partial or full builds from the commit messages...

Getting into a bit of work but I think update churn is bad enough that its worth it.

**Comment by HugLifeTiZ** [comment]
Do you remove your old drivers from your flatpak installation when you upgrade to a new driver series? I don't think `flatpak uninstall --unused` is smart enough to remove those, so you have to look for your unused nvidia versions and remove them manually, but at the very least, restricting the update churn to the driver series you're actually using should help a lot.

**Comment by Flashwalker** [comment]
Still have this ussie in 1.14.0 on Ubuntu 22.04.
It literally downloads nvidia driver on every install. Even the same as installed too.

**Comment by guihkx** [comment]
Someone can correct me if I'm wrong, but you can probably prevent a re-install by simply 'masking' the drivers, e.g.

```
flatpak mask org.freedesktop.Platform.GL.nvidia-470-141-03 org.freedesktop.Platform.GL32.nvidia-470-141-03
```

You check your masked apps/runtimes with:

```
flatpak mask
```

**Comment by HugLifeTiZ** [comment]
I think this will be a good workaround because it will prevent the redundant updates that come from the sane nvidia driver being rebuilt for no reason just because another drier got added. However this could become problematic in the future should an actual change happen to the Flathub package that needs to be downloaded for an already-installed driver. Nvidia updates from Flathub are definitely very "cry wolf" until this bug gets fixed for real.

It's been over two years since this issue was opened.

It's the build.sh script that builds every version? How does it get called? What sort of state and/or data persists between builds?

**Comment by barthalion** [comment]
> It's the build.sh script that builds every version?

Yes.

> How does it get called?

See arguments assignments at https://github.com/flathub/org.freedesktop.Platform.GL.nvidia/blob/master/build.sh#L5

> What sort of state and/or data persists between builds?

None. You can query available refs with flatpak CLI in build.sh.

**Comment by biji** [comment]
Probably same downloaded driver can be cached?

**Comment by Flashwalker** [comment]
Is there any way to prevent the nvidia driver from updating at all, as long as a fresh version is not required as dep?

**Comment by TingPing** [comment]
@Flashwalker `flatpak mask $PACKAGE_NAME`

**Comment by joezhouchenye** [comment]
Why on earth would flatpak install Nvidia drivers? Using Ubuntu 22.04, two software stores? FIne. Now, this? I even cannot update Nvidia drivers from the system app software and update anymore and all is greyed. Consider uninstalling fatpak.

**Comment by TingPing** [comment]
@joezhouchenye It doesn't have to. [Ubuntu could package the drivers for flatpak.](https://blog.tingping.se/2018/08/26/flatpak-host-extensions.html).

**Comment by youainti** [comment]
Having read through this and found it helpful, I just wanted to summarize the step I took to fix it.

```bash
flatpak uninstall --unused
flatpak list --runtime
```
Look for the nvidia version in use. I got something along the lines of:

```
Name                                              Application ID                                            Version              Branch                  Installation
Freedesktop Platform                              org.freedesktop.Platform                                  22.08.7              22.08                   system
Mesa                                              org.freedesktop.Platform.GL.default                       21.3.9               21.08                   system
Mesa                                              org.freedesktop.Platform.GL.default                       22.3.4               22.08                   system
Mesa (Extra)                                      org.freedesktop.Platform.GL.default                       22.3.4               22.08-extra             system
nvidia-470-161-03                                 org.freedesktop.Platform.GL.nvidia-470-161-03                                  1.4                     system
ffmpeg-full                                       org.freedesktop.Platform.ffmpeg-full                                           21.08                   system
openh264                                          org.freedesktop.Platform.openh264                         2.1.0                2.0                     system
openh264                                          org.freedesktop.Platform.openh264                         2.1.0                2.2.0                   system
Freedesktop SDK                                   org.freedesktop.Sdk                                       21.08.12             21.08                   system
GNOME Application Platform version 41             org.gnome.Platform                                                             41                      system
```
you are going to use this line: 
```
nvidia-470-161-03                                 org.freedesktop.Platform.GL.nvidia-470-161-03                                  1.4                     system
```
You then mask the appropriate packages
```
flatpak mask  org.freedesktop.Platform.GL.nvidia-470-161-03
```
now run your updates
```
flatpak update package.name.goes.here
```

**Comment by biji** [comment]
It seems repeated download happen only after updating kernel

**Comment by zaratul00** [comment]
> It seems repeated download happen only after updating kernel

And thats not fun when you use Arch ..

**Comment by evdcush** [comment]
Same problem.
`flatpak mask` hack worked.

**Comment by poltpolt** [comment]
> Same problem. `flatpak mask` hack worked.

how did you do it?
i did 
➜  ~ flatpak mask -u org.freedesktop.Platform.nvidia
➜  ~ flatpak mask  org.freedesktop.Platform.nvidia 
with no luck - still get them listed for update

**Comment by poltpolt** [comment]
so it is not a mask but in fact an exclusion for particular package... for me a mask is something that covers more then one particular identifier but a bit broader

**Comment by guihkx** [comment]
Well, you could technically do something like `flatpak mask "org.freedesktop.Platform.GL.nvidia-*"`, but if e.g. you upgrade your host drivers to 550.02.01, you'll have to install the Flatpak package for this new version *manually*, because `flatpak update` won't install it automatically for you anymore.

**Comment by Flashwalker** [comment]
I wrote a script you want to run after you upgrade your host system nvidia driver:

```bash fupdate-nvidia-driver
#!/bin/bash -e

oldver="$(flatpak list | sed -rn 's/org.freedesktop.Platform.GL.nvidia-([-[:digit:]]+)[[:space:]]+.*$/\1/p' | sed -rn 's/^.*[[:space:]]+(.*)/\1/p')"
newver="$(cat /sys/module/nvidia/version | tr '.' '-')"

function echover {
    echo "flatpak nvidia driver:  ${oldver}"
    echo "host nvidia driver:     ${newver}"
}

# If host driver version older than flatpak one
if [ ${newver//-/} -gt ${oldver//-/} ]; then
    echover
    echo 'Updating...'
    sudo flatpak install org.freedesktop.Platform.GL.nvidia-"${newver}" &&
    sudo flatpak mask org.freedesktop.Platform.GL{,32}.nvidia-"${newver}"
    sudo flatpak mask --remove org.freedesktop.Platform.GL{,32}.nvidia-"${oldver}"
    sudo flatpak remove org.freedesktop.Platform.GL{,32}.nvidia-"${oldver}"
    exit
# If both tha same
elif [ ${newver//-/} -eq ${oldver//-/} ]; then
    echover
    echo 'same. do nothing.'
    exit
else
    echover
    exit
fi

```

**Comment by rambo919** [comment]
> fupdate-nvidia-driver.sh

This is helpful but results in "unary operator expected" errors in lines 12 and 21. Easily fixed by double brackets instead of single one's.

**Comment by Flashwalker** [comment]
> This is helpful but results in "unary operator expected" errors in lines 12 and 21. Easily fixed by double brackets instead of single one's.

Thnx, fixed


> However it then ends in "error: No current masked pattern matching org.freedesktop.Platform.GL.nvidia-" which I assume is part of the removal part of the previous runtime.

Perhaps regexp for oldver is not working for you. Try running it and see if it works:  
```bash
flatpak list | sed -rn 's/org.freedesktop.Platform.GL.nvidia-([-[:digit:]]+)[[:space:]]+.*$/\1/p' | sed -rn 's/^.*[[:space:]]+([-[:digit:]]+)/\1/p'
```
output should be like: 545-29-06

**Comment by rambo919** [comment]
sigh.... the morons did it again and released a hotfix version willy nilly and now no flatpaks will run anymore and once again I probably have to wait a month for someone to fix it.

Output is nvidia-550-54-14 which is the most current flatpak version..... driver version is nvidia-550-54-15. However runtimes also also installed are nvidia-525-125-06, nvidia-525-147-05 and nvidia-545-23-08. For some reason DDM refuses to uninstall 525 so technically I have multiple drivers installed which in itself confuses the system.... I also cannot manually downgrade or uninstall 525 because of dependency problems which is a separate issue.

**Comment by Flashwalker** [comment]
@rambo919 
I just ran `fremove org.freedesktop.Platform.GL.nvidia-OLD_VERSION` and it was successfully removed. I just ignored the dependency prompt and forced the removal.

**Comment by rambo919** [comment]
> @rambo919 I just ran `fremove org.freedesktop.Platform.GL.nvidia-OLD_VERSION` and it was successfully removed. I just ignored the dependency prompt and forced the removal.

No I mean on system level 525 is still installed as well as the latest one the corresponding flatpaks. I blame nvidia for simply not caring about anyone other than AI farmers at this point. Unlike in windows you cannot downgrade the driver if you need to if you use the nvidia repo, you are forced to upgrade unless you remember to force freeze the version.

It took me far too long to realize what their moron ceo meant by "buying more and more often is buying cheaper".... he obviously considers the common consumer an irritation at this point.

But back to the point, people have commented that if they force remove old driver flatpak runtimes.... they just automatically reinstall next time.

**Comment by TingPing** [comment]
flatpak only auto installs the version that matches your host.

The next major release of flatpak will also auto-remove versions that dont match.

**Comment by Flashwalker** [comment]
> > @rambo919 I just ran `fremove org.freedesktop.Platform.GL.nvidia-OLD_VERSION` and it was successfully removed. I just ignored the dependency prompt and forced the removal.
> 
> No I mean on system level 525 is still installed as well as the latest one the corresponding flatpaks. I blame nvidia for simply not caring about anyone other than AI farmers at this point. Unlike in windows you cannot downgrade the driver if you need to if you use the nvidia repo, you are forced to upgrade unless you remember to force freeze the version.
> 
> It took me far too long to realize what their moron ceo meant by "buying more and more often is buying cheaper".... he obviously considers the common consumer an irritation at this point.
> 
> But back to the point, people have commented that if they force remove old driver flatpak runtimes.... they just automatically reinstall next time.

You can run this  
to install e.g. all nvidia 550
```bash
sudo apt-get install nvidia-compute-utils-550 nvidia-dkms-550 nvidia-kernel-common-550 nvidia-driver-550 nvidia-utils-550 xserver-xorg-video-nvidia-550 libnvidia-gl-550 libnvidia-compute-550 libnvidia-decode-550 libnvidia-encode-550 libnvidia-extra-550 libnvidia-cfg1-550 libnvidia-fbc1-550 libnvidia-common-550 nvidia-firmware-550-* libnvidia-gl-550:i386 libnvidia-compute-550:i386 libnvidia-decode-550:i386 libnvidia-encode-550:i386 libnvidia-fbc1-550:i386 libnvidia-compute-550 nvidia-compute-utils-550 nvidia-dkms-550 nvidia-kernel-common-550
```

and remove all nvidia 525
```bash
sudo apt-get purge nvidia-compute-utils-525 nvidia-dkms-525 nvidia-kernel-common-525 nvidia-driver-525 nvidia-utils-525 xserver-xorg-video-nvidia-525 libnvidia-gl-525 libnvidia-compute-525 libnvidia-decode-525 libnvidia-encode-525 libnvidia-extra-525 libnvidia-cfg1-525 libnvidia-fbc1-525 libnvidia-common-525 nvidia-firmware-525-* libnvidia-gl-525:i386 libnvidia-compute-525:i386 libnvidia-decode-525:i386 libnvidia-encode-525:i386 libnvidia-fbc1-525:i386 libnvidia-compute-525 nvidia-compute-utils-525 nvidia-dkms-525 nvidia-kernel-common-525
```

**Comment by rambo919** [comment]
I tried something like that when i tried to force downgrade had to do a timeshift because I lost the ability to boot.... those specific packages don't actually show up for me in apt anyway.... something to do with how MX Linux tweaks Debian probably or just a Debian problem I dunno.

Though oddly now that I checked a few all of the old driver flatpaks complain that heroic launcher and only heroic launcher needs them..... using Warehouse uninstalled that, then the old driver flatpaks.....  now it does not redownload them. I still have the problem of the mystery dual driver setup system side but it seems like the main problem flatpak wise might have been heroic blocking things.

At least this time the flatpak driver hotfix issue was solved quicker this time and it updated to the right one today.... which bodes well for the future.

**Comment by TingPing** [comment]
> only heroic launcher needs them

This package using the 32bit drivers (note `GL32`) but it shouldn't be a dependency like that I'd think.

**Comment by rambo919** [comment]
It can work but it's a bit flakey.

**Comment by FlameSoulis** [comment]
This is very much still a thing. Seriously, why is this the ONLY package that takes a million years to download?

For the record, the other packages usually download at about 30MB/s. This is the ONLY one that is unusually slow, and sometimes goes as slow as bytes/s.
![image](https://github.com/user-attachments/assets/c3db3296-d5ec-4d56-a094-351c177f7d13)

**Comment by guihkx** [comment]
@FlameSoulis Flatpak downloads the driver straight from NVIDIA servers, while most apps are downloaded from Flathub servers. That's why you're seeing that difference in download speeds.

Though I personally never had slow speeds while downloading from NVIDIA servers either, so YMMV.
